CMakeLists.txt

```
@@ -81,7 +85,12 @@ endif()
 enable_testing(true)
 
 # SPIR-V headers
-add_subdirectory(externals/SPIRV-Headers EXCLUDE_FROM_ALL)
+if (SIRIT_USE_SYSTEM_SPIRV_HEADERS)
+    find_package(SPIRV-Headers REQUIRED)
+    add_library(SPIRV-Headers ALIAS SPIRV-Headers::SPIRV-Headers)
+else()
+    add_subdirectory(externals/SPIRV-Headers EXCLUDE_FROM_ALL)
+endif()
```

As always, I would prefer this to be

```suggestion
find_package(SPIRV-Headers)
if (NOT SPIRV-Headers_FOUND)
    add_subdirectory(externals/SPIRV-Headers EXCLUDE_FROM_ALL)
    add_library(SPIRV-Headers::SPIRV-Headers ALIAS SPIRV-Headers)
endif()
```

so that if you already have SPIRV-Headers installed things just work

---

It should be possible to NOT use the system library even if presents, in case of incompatibilities.

---

I'm not convinced... If the found headers are not compatible `find_package()` should mark the dependency as not found and fallback to the bundled one, it's not up to the user to check if their headers are compatible or not.

And even then, the option should be opt out, not opt in, i.e. if I have the headers installed they will be used automatically, but if I know they are not compatible I can exclude them by disabling the option.

Making the build fail without a fallback isn't that useful either. If you can fallback to a bundled dependency, why would you want the build to fail?

---

One example is if you make a package and the build silently fallback to a bundled library, you end up with a package dependent on a now useless system dependency without knowing it.

And the option is off by default to comply with the default behavior of the code before this PR.

---

> One example is if you make a package and the build silently fallback to a bundled library, you end up with a package dependent on a now useless system dependency without knowing it.

Uhm, how can that happen? I don't know about arch, but in Debian library dependencies are automatically extracted by looking at the symbols used

---

In Arch the maintainer manually declare the dependencies in the [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=yuzu) :p

---

I see, thanks. Maybe I'm just biased towards this "automatically use system deps if available, fallback to the bundled subprojects otherwise" approach because I really enjoy using [Meson](https://mesonbuild.com), but I still prefer having system deps used without having to look at all the various CMakeLists.txt files to figure out the correct option to enable :)

---

> Making the build fail without a fallback isn't that useful either. If you can fallback to a bundled dependency, why would you want the build to fail?

> One example is if you make a package and the build silently fallback to a bundled library, you end up with a package dependent on a now useless system dependency without knowing it.

I quite agree... Meson makes this work by having `--wrap-mode=nofallback` as a builtin option that makes it never fall back to a bundled dependency, and there are never useless system dependencies in a distro package because the package simply errors out when necessary preconditions are not met.

> Maybe I'm just biased towards this "automatically use system deps if available, fallback to the bundled subprojects otherwise" approach because I really enjoy using [Meson](https://mesonbuild.com/),

Meson is indeed great, because it gives both the upstream authors and the downstream builders, ***choice***.

One of those choices does what you want -- automatically fall back to the bundled subprojects.

One of those choices does what @abouvier wants -- error out if you asked for system dependencies but couldn't find them.

Meson is *not* a good example of automatically making decisions the user is unable to override though, which is what you seem to be suggesting...

> but I still prefer having system deps used without having to look at all the various CMakeLists.txt files to figure out the correct option to enable :)

This could be fixed if CMake added functionality for showing options akin to autotools `./configure --help`, but I guess that will never happen. A shame.

---

> Meson is _not_ a good example of automatically making decisions the user is unable to override though, which is what you seem to be suggesting...

I mean that Meson has good defaults, but of course choice is important. I wanted this to be opt-out, not opt-in, but whatever :)

> This could be fixed if CMake added functionality for showing options akin to autotools `./configure --help`, but I guess that will never happen. A shame.

I seem to remember that CMake GUI shows all the options and lets you toggle them, so that's definitely possible.

Anyway... I'm curious now. Do you have an interplanetary radar that notifies you whenever someone says the word "Meson"?

---

> I wanted this to be opt-out, not opt-in, but whatever :)

That could be simply done by changing the default value of the option without changing the resolution method to one that does fallbacks.

The problem is that doing fallback resolution properly is non-trivial, it's best if the framework handles this for you, but cmake doesn't.

> I seem to remember that CMake GUI shows all the options and lets you toggle them, so that's definitely possible.

Depending on the GUI is not an answer.

Anyway editing the cache variables doesn't actually do what's wanted. You see many things that aren't options, and cmake happily allows you to use `-D` switches that are neither options nor cache variables to have the *effect* of options. Also, editing the cache is only possible IIRC after you have configured, possibly unsuccessfully.

The best compromise is probably having a section of the README that documents the possible options and allowed values.

> Anyway... I'm curious now. Do you have an interplanetary radar that notifies you whenever someone says the word "Meson"?

It's not interplanetary, lol, I wish.